### DISCUSSION TITLE
IaC framework decision

### DISCUSSION THREAD

**Discussion by fgmadeira** [diconium/ethereal-nexus#76] [Polls]
Nexus will be a self-hosted tool that should be easy to install for any user and project.

For this, we should have an easy IaC option for installation. There are several options, and we need to decide which one to use. Below are some of the most common IaC tools that are not proprietary to any specific cloud provider:

[Terraform](https://www.terraform.io/): An open-source tool by HashiCorp that enables users to define and provision infrastructure using a high-level configuration language. Terraform supports multiple cloud providers and on-premises environments.

[Ansible](https://www.ansible.com/): An open-source automation tool by Red Hat that allows users to automate software provisioning, configuration management, and application deployment. Ansible uses YAML-based playbooks to define automation tasks.

[Pulumi](https://www.pulumi.com/): An open-source IaC tool that allows users to define infrastructure using general-purpose programming languages like TypeScript, JavaScript, Python, Go, and .NET.

[OpenTofu](https://opentofu.org/): An open-source alternative to Terraform, OpenTofu allows for defining and provisioning infrastructure in a similar manner, ensuring portability across different cloud providers and on-premises environments.

Each of these tools has its own strengths and best-use scenarios, and we should evaluate them based on our specific requirements and environment. Let's discuss these options and decide which one will best suit our needs for the Nexus project.

If you know of other IaC tools that could be beneficial for us to consider, please feel free to vote on the 'Other' and leave a comment with your idea. 

We are open to exploring all viable options to ensure we choose the best tool for our needs.

**EDIT 01-07-2024**: Poll will be closed by 04-07-2024 so it can be planned in the next feature planning meeting.

**Comment by joaotab** [comment]
Maybe something we need to consider about Terraform vs OpenTofu is that they recently changed their license model to BSL which is technically not opensource and might now go against company policies (looking at you Whitesource, Blackduck).

https://spacelift.io/blog/terraform-license-change

**Comment by joaotab** [comment]
To add some Pros&Cons:

Terraform
➕ The de-facto standard, widely supported and documented
➖ Hard to customize, e.g. for private clouds with limited permissions
➖ Not opensource anymore

OpenTofu
➕ opensource fork of Terraform
➖ same

Ansible
➕ Wide array of plug-ins and very flexible to customize
➖ Learning curve; much lower level tool
➖ Becomes very slow when scaling

Pulumi
➕ More modern and higher level tool; Supports multiple languages already known to devs
➖ Documentation / AI tool has some gaps; working examples harder to find

**Comment by fgmadeira** [comment]
This discussion is now closed! 

The winner seems to be Pulumi, we will now take this decision into implementation of the IaC of ethereal nexus environments.
Thanks to all that voted!
